### DISCUSSION TITLE
Microphone denoise

### DISCUSSION THREAD

**Discussion by alexbelgium** [Nachtzuster/BirdNET-Pi#74] [General]
Hi, I see that several apps exist to denoise microphones such as noisetorch ; rnnoise ; easyeffects, has anyone has experience on using those as a way to improve microphone sensitivity and reduce false positives? thanks!

Edit : a discussion was with similar ideas on https://github.com/mcguirepr89/BirdNET-Pi/discussions/253 ; from the gist of it it seems that it would not improve detection ?

Edit2 : playing with alsamixer settings actually seem to provide already a good mix between signal and noise

**Comment by krummrey** [comment]
From what I got from this [BirdNET-Pi/discussions/39](https://github.com/mcguirepr89/BirdNET-Pi/discussions/39) is, that noise does not throw the detection off. So while it might be pleasing for the ears to use a ultra low noise microphone or to filter out the noise, it doesn't really help the detection algorithm.

**Comment by lloydbayley** [comment]
I thought about this in the initial stages of my running Birdnet-Pi and tried various alsamixer settings, foam mic covers and indeed 'deadcat'-type covers. I didn't go with software solutions, as I was concerned that it might reduce recognition as some of the calls are extremely similar and could be lumped into the same 'detection'. Ultimately, wind, barking dogs, kids screaming and trucks reversing beeps were all mostly 'screenable' by excluding certain birds that were detected as the previous things. I was very lucky, in that, the birds I needed to exclude are not birds that frequent my local area. 

Whilst I've had overbearing noises that will change or indeed in some cases preclude a valid detection, there are plenty of other times that the detection is correct, thus sating my penchant for precision. You know how it is - "What you lose on the swings you gain on the roundabouts"!

It's a jolly-good question about how best to manage false-detections. The 'noise' coming in is still perfectly 'analysable'  but we will always have to filter out things that are just not what we want. Ie, barking dogs and backup beeps etc. It will vary hugely around the globe as certain regions will never have a bird that matches a backup beep, whilst others will. I suspect all regions will have at least ONE bird that might be detected when a dog barks, as that is Murphy's law! :)

**Comment by alexbelgium** [comment]
thanks for the replies !

Indeed currently I'm managing false positives with both exclusion, as you suggested, and I've also made a "conversion feature" on my addon (that for examples, converts every Falco subbuteo detected as a Falco tinnunculus as clearly they did not include young Falco tinnunculus in their models). I've seen at least 3 birds where their calls are systematically mistaken for another birds' song in a quite obvious way (and most importantly repeatable).

Indeed I wanted to have the most "pleasing" sound as well as most efficient ; and my newly installed lavalier microphone provides very low noise ; but very low volume too - by amplifying the volume beyond the Alsamixer's 100% I managed to have clearer songs but then noise increased in a linear manner to signal (a logarithmic amplifier would probably be better to give more importance to signals over noise). And therefore I was looking for the denoiser :-)

But indeed based both your feedbacks I'll have to just be happy that the app is working greatly to identify the birds, which is the most important !

**Reply by lloydbayley** [reply]
Interesting that you were able to clear it up beyond 100%. That's great. It's a bit of a limitation when you get clearer sounds but more 'rubbish' with a higher gain.

I too, have a bird here which had developed his/her own unique variation on the traditional call, and therefore is not detected as the appropriate species.

There's not a lot I can do about that at this stage of development as it would require a training extension I suppose....

I'm off to make a pot of tea. Who wants one?

**Reply by alexbelgium** [reply]
Subscribing to some tea ! 3pm here, so perfect timing. Will help fuel other "interesting ideas" on how to further tweak the systems until I run out of things to try and just accept that Birdnet-pi works great as it is :-)

**Reply by alexbelgium** [reply]
After some testing, enabling pulseaudio echo cancel module does seem to increase detection confidences!

**Reply by lloydbayley** [reply]
How interesting. It's quite the trial and error technique sometimes...

**Reply by alexbelgium** [reply]
I had an interesting observation : my pulseaudio server reset this morning and lost both the gain and noise cancellation features. The spectrogram was nearly black but it still gave acceptable detectabilities. Here is what a spectrogram looks with minimal gain, gain at 100%, and gain at 100% + noise cancelling. I think the increased detection values are only valid when the source is quite near to the microphone as the algorithm seems to amplify sounds that are already above background ; while without noise cancelling there is better sensitivity for very low volume noise

![image](https://github.com/Nachtzuster/BirdNET-Pi/assets/44178713/a5a142fb-43d3-4f1a-b919-b9d79ee2cedc)

**Reply by lloydbayley** [reply]
Yeh, that's REALLY weird. I've never had a spectrogram that black. I'm quite shocked that the detections continued! It's a funny old thing, isn't it?

**Reply by alexbelgium** [reply]
I thought the audio was shutdown and didn't understand why it was still detecting birds until I turned the light way up on my phone ;-)

I found a way to play with it in a more objective way : generate wav files then test them with the Birdnet-analyzer desktop app - it is quite informative ! For example, even though the model is in theory trained between 150hz and 150000hz (according to its changelist), the detection quality decreases if a highpass filter is applied to suppress frequencies below 150hz and then analyse in the app!

Edit : latest version of the model is trained from 0 to 15 khz

**Comment by Frono** [comment]
I wish I could see [this article](https://ieeexplore.ieee.org/document/9793420/references#references)... Not that I could understand how to implement, but the abstract reads like one of my objectives...

**Reply by dalogue1** [reply]
[A_Birdsong_Denoising_using_Cascaded_Combination_of_Phase_Spectrum_Compensation_and_Wavelets.pdf](https://github.com/user-attachments/files/16018746/A_Birdsong_Denoising_using_Cascaded_Combination_of_Phase_Spectrum_Compensation_and_Wavelets.pdf)

**Reply by Frono** [reply]
Thank you!

**Comment by MichaelPWalter** [comment]
Expanding on the denoising, has someone tried MEMs-based microphones? From my limited understanding, the digital communication protocols used by those devices (PDM, I2S) should be less susceptible to noise on longer cable runs.

**Reply by alexbelgium** [reply]
Hi, from what I understood only analog signal (aux cable) is susceptible to noise induced by cable length. So if you convert it to rtsp (using an rpi attached to a POE ethernet cable) or to usb (using USB extenders that then send the signal on a ethernet cable) you're safe.

The biggest source of noise I found on my system was the power supply (same as when pc speakers buzz, bought a special usb thing to remove this) ; the mic ; and the aux to usb converter.

**Reply by MichaelPWalter** [reply]
Thats what I imagined. I have the worst possible setup right now: RPi02W - microUSB to USB - USB to 3.5 mm - and a 6 feet cable to a lavalier mic. The longest cable run is also the most vulnerable. Im hoping to upgrade to a MEMs mic so the analog signal only travels on the breakout board and the long travel back to the RPi will be digital. I have not thought about the power supply yet. Might be worth checking with my oscilloscope.

**Reply by alexbelgium** [reply]
Might not be so bad - @lloydbayley has introduced me to [USB extenders](https://www.amazon.com/Siawclub-Ethernet-Compatible-Computers-Keyboards/dp/B0CTGTH26N) that are great. For less than 20€/$ you could have this plus a small ethernet cable to replace it. I've also added a [USB denoiser](https://www.amazon.com/DEVMO-USB-ADUM3160-Isolation-Protection/dp/B09L4QJM14) to remove power supply noise but not sure if it really helped. I've tried [ferrite beads](https://www.amazon.com/20pcs-Clip-Ferrite-Suppressor-Diameter/dp/B0CSXZWGD6) as in theory it should server the same purpose. Perhaps you could post a night spectrogram to see what your noise looks like but in theory it shouldn't be that bad. Watcht-out that noise can also just come from the environment (electricity humming...)
